sdk/iot/core/inc/az_iot_hub_client.h

```diff
@@ -0,0 +1,505 @@
+ *                          filter.
+ * @param out_mqtt_topic_filter The output #az_span containing the MQTT topic filter.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_hub_client_methods_subscribe_topic_filter_get(
+    az_iot_hub_client const* client,
+    az_span mqtt_topic_filter,
+    az_span* out_mqtt_topic_filter);
+
+/**
+ * @brief A method request received from IoT Hub.
+ *
+ */
+typedef struct az_iot_hub_client_method_request
+{
+  az_span request_id; /** @brief The request id.
+                       * @note The application must match the method request and method response. */
+  az_span name; /** @brief The method name. */
+} az_iot_hub_client_method_request;
+
+/**
+ * @brief Attempts to parse a received message's topic.
+ *
+ * @param client
+ * @param received_topic An #az_span containing the received topic.
+ * @param out_request If the message is a method request, this will contain the
+ *                    #az_iot_hub_client_method_request.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_hub_client_methods_received_topic_parse(
+    az_iot_hub_client const* client,
+    az_span received_topic,
+    az_iot_hub_client_method_request* out_request);
+
+/**
+ * @brief Gets the MQTT topic that must be used to respond to method requests.
+ *
+ * @param client
+ * @param request_id The request id. Must match a received #az_iot_hub_client_method_request
+ *                   request_id.
+ * @param status The status. (E.g. 200 for success.)
+ * @param mqtt_topic An empty #az_span with sufficient capacity to hold the MQTT topic.
+ * @param out_mqtt_topic The output #az_span containing the MQTT topic.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_hub_client_methods_response_publish_topic_get(
+    az_iot_hub_client const* client,
+    az_span request_id,
+    uint16_t status,
+    az_span mqtt_topic,
+    az_span* out_mqtt_topic);
+
+/**
+ *
+ * Twin APIs
+ *
+ */
+
+/**
+ * @brief Azure IoT Hub status codes.
+ *
+ */
+typedef enum
+{
+  // Service success codes
+  AZ_IOT_HUB_CLIENT_STATUS_OK = 200,
+  AZ_IOT_HUB_CLIENT_STATUS_ACCEPTED = 202,
+  AZ_IOT_HUB_CLIENT_STATUS_NO_CONTENT = 204,
+
+  // Service error codes
+  AZ_IOT_HUB_CLIENT_STATUS_BAD_REQUEST = 400,
+  AZ_IOT_HUB_CLIENT_STATUS_UNAUTHORIZED = 401,
+  AZ_IOT_HUB_CLIENT_STATUS_FORBIDDEN = 403,
+  AZ_IOT_HUB_CLIENT_STATUS_NOT_FOUND = 404,
+  AZ_IOT_HUB_CLIENT_STATUS_NOT_ALLOWED = 405,
+  AZ_IOT_HUB_CLIENT_STATUS_NOT_CONFLICT = 409,
+  AZ_IOT_HUB_CLIENT_STATUS_PRECONDITION_FAILED = 412,
+  AZ_IOT_HUB_CLIENT_STATUS_REQUEST_TOO_LARGE = 413,
+  AZ_IOT_HUB_CLIENT_STATUS_UNSUPPORTED_TYPE = 415,
+  AZ_IOT_HUB_CLIENT_STATUS_THROTTLED = 429,
+  AZ_IOT_HUB_CLIENT_STATUS_CLIENT_CLOSED = 499,
+  AZ_IOT_HUB_CLIENT_STATUS_SERVER_ERROR = 500,
+  AZ_IOT_HUB_CLIENT_STATUS_BAD_GATEWAY = 502,
+  AZ_IOT_HUB_CLIENT_STATUS_SERVICE_UNAVAILABLE = 503,
+  AZ_IOT_HUB_CLIENT_STATUS_TIMEOUT = 504,
+} az_iot_hub_client_status;

```

Comment:
>az_iot_hub_client_status [](start = 2, length = 24)

What happens if we get a status code from IoThub that is not explicitly defined in this range?

Reply:
I thought about 2 scenarios:

1. Apps may choose to use the enums. In that case the app will likely not be able to handle the response and crash. That would be a service breaking change.
1. Apps will use the provided `az_iot_hub_client_is_success_status` and `az_iot_hub_client_is_retriable_status` APIs. Those should be implemented using ranges instead of actual codes and so the apps should work without any changes.

Reply:
1 - What happens if we go through a proxy (or a load balancer or a ... IoTHub doesn't control) at some point?  Could the proxy return an error code outside the error range in the enum list?
2 - Even if 1 isn't an issue, is service team onboard with fact that these errors are a strict contract and they can't add new ones?

Reply:
Or actually I may be thinking too HTTP in my comment above, since this is ultimately MQTT land and any sort of proxy error would be a web error at the client's transport layer, correct?

Reply:
Right: the md file is setting clear expectations on us not supporting WebSocket at this layer.
That said, Edge is a reasonable alternative that would enable 1. 
Still my thinking is: apps that have code testing for a particular value probably only want to address that particular value. 
Most apps should rely on the two APIs I mention. Those should be implemented to cover the entire uint16_t range, not just 100-500s.